### PR TITLE
Closes #2576 add query string to bypass WP Rocket optimizations

### LINKED ISSUE #2576: Create dedicated query string to bypass WP Rocket
Support requires a way to bypass WP Rocket, for debugging and comparison purposes. 
[Since we are now going to again start mixing up which features are applied on uncached pages,](https://github.com/wp-media/wp-rocket/issues/2543#issuecomment-618380044) let's create a dedicated query string that we know will bypass ALL of WP Rocket's features. 

We are aware that some have to be applied on uncached pages for technical reasons:
https://docs.wp-rocket.me/article/1285-bypass-wp-rockets-caching-and-optimizations

But besides those, the new query string will turn off everything else including anything that has been chosen to be applied on an uncached page. 

Secondarily, if possible, a dedicated query string that bypasses _only_ RocketCDN would be valuable when doing debugging on that specific feature. (props @camilamadronero )

### PR DESCRIPTION
**Pull request by iCaspar**
Adds the query string bypass feature: closes #2576

### Code changes

- Add a new function to the API: `rocket_bypass()` checking for query string `?nowprocket` in current url.
- Add conditional bail-outs to processes to be short-circuited by the query string.
```
    Conversion of dynamic CSS/JS files to static files
    Disable Emoji
    Disable WordPress embeds
    Heartbeat settings
    Preload fonts
    WooCommerce Empty Cart Caching
```

Since we're touching old code, we're also taking this opportunity to move it to the new architecture, details below:

- [x] Create a `Engine\HeartBeat` folder, move & rename the subscriber from `inc/classes/subscriber/class-heartbeat-subscriber`
- [x] Create a `Engine\HeartBeat\ServiceProvider` to move the instantiation of the class from `inc/classes/ServiceProvider/class-common-subscribers.php` to this one
- [x] Create a `Engine\Media\Embeds` folder, take the code from `inc/common/embeds.php` and transform it into a subscriber
- [x] Create a `Engine\Media\Emojis` folder,  take the code from `inc/common/emojis.php` and transform it into a subscriber
- [x] Update the `Engine\Media\ServiceProvider` to load the 2 new classes
- [x] Update `Plugin` to register the new ServiceProviders, and the new subscribers
- [x] Deprecate the old functions, create a class_alias for `class-heartbeat-subscriber`

### Acceptance criteria
- When using the querystring `nowprocket` in the URL (http://example.org/?nowprocket=1), there should be no optimization applied to the page (should be the same as the no cache page)
- This includes the following optimizations, that were previously applied also on uncached pages:
```
Conversion of dynamic CSS/JS files to static files
Disable Emoji
Disable WordPress embeds
Heartbeat settings
Preload fonts
WooCommerce Empty Cart
```

As an aside, the following features have been migrated to the new code architecture, so a test to check that there is no regression is welcome:
- Disable Emoji
- Disable WordPress embeds
- Heartbeat settings

Automated tests have been added for the emojis & embeds.

### CODE REVIEW

**Comment by iCaspar** [comment]
3 erring unit tests as of here need to be updated to stub the new function.

File: inc/Engine/Heartbeat/ServiceProvider.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	 *
+	 * @since 3.6
+	 *
+	 * @return void
+	 */
+	public function register() {
+		$options = $this->getContainer()->get( 'options' );
+
+		$this->getContainer()->share( 'heartbeat_subscriber', 'WP_Rocket\Engine\Heartbeat\HeartbeatSubscriber' )
+			->withArgument( $options );
```
Comment (Tabrisrp) [inline inc/Engine/Heartbeat/ServiceProvider.php:37]:
```suggestion
		$this->getContainer()->share( 'heartbeat_subscriber', 'WP_Rocket\Engine\Heartbeat\HeartbeatSubscriber' )
			->withArgument( $this->getContainer()->get( 'options' ) );
```

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 2]
```diff
@@ -0,0 +1,237 @@
+<?php
+namespace WP_Rocket\Engine\Media\Embeds;
+
+use WP_Rocket\Admin\Options_Data;
+use WP_Rocket\Event_Management\Subscriber_Interface;
+use WP_Rocket\ThirdParty\ReturnTypesTrait;
+use function rocket_bypass;
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:7]: Is the import required?

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+	 */
+	public function __construct( Options_Data $options ) {
+		$this->options = $options;
+	}
+
+	/**
+	 * Return an array of events that this subscriber wants to listen to.
+	 *
+	 * @since  3.7
+	 * @access public
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:38]:
```suggestion
```

We don't need the access in the docblock, it's already indicated by the `public` keyword in the method declaration

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+	 *
+	 * @since 3.7 Moved to new architecture.
+	 * @since 3.3.3
+	 *
+	 * @param array $data The response data.
+	 *
+	 * @return array|false Response data or false if in a REST API context.
+	 */
+	public function disable_embeds_filter_oembed_response_data( $data ) {
+		if ( defined( 'REST_REQUEST' ) && REST_REQUEST ) {
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:165]:
```suggestion
		if ( rocket_get_constant( 'REST_REQUEST' ) ) {
```

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+	 * This is used to unregister the `core-embed/wordpress` block type.
+	 */
+	public function disable_embeds_enqueue_block_editor_assets() {
+		if ( ! $this->can_embed() ) {
+			return;
+		}
+
+		wp_enqueue_script(
+			'rocket-disable-embeds',
+			WP_ROCKET_ASSETS_JS_URL . 'editor/editor.js',
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:191]:
```suggestion
			rocket_get_constant( 'WP_ROCKET_ASSETS_JS_URL' ) . 'editor/editor.js',
```

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+			[
+				'wp-edit-post',
+				'wp-editor',
+				'wp-dom',
+			],
+			WP_ROCKET_VERSION,
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:197]:
```suggestion
			rocket_get_constant( 'WP_ROCKET_VERSION' ),
```

File: inc/Engine/Media/Embeds/EmbedsSubscriber.php [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+	 * @since 3.7
+	 *
+	 * @return bool
+	 */
+	private function can_embed() {
+		if ( rocket_bypass() || ! $this->options->get( 'embeds' ) ) {
+			return false;
+		}
+
+		return true;
```
Comment (Tabrisrp) [inline inc/Engine/Media/Embeds/EmbedsSubscriber.php:235]: We can simplify this as a one liner

```suggestion
		return ( ! rocket_bypass() || (bool) $this->options->get( 'embeds', 0 ) );
```

File: inc/Engine/Media/Emojis/EmojisSubscriber.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	 */
+	public function __construct( Options_Data $options ) {
+		$this->options = $options;
+	}
+
+	/**
+	 * Return an array of events that this subscriber wants to listen to.
+	 *
+	 * @since  3.7
+	 * @access public
```
Comment (Tabrisrp) [inline inc/Engine/Media/Emojis/EmojisSubscriber.php:34]:
```suggestion
```

File: inc/Engine/Media/Emojis/EmojisSubscriber.php [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	 * @since 3.7
+	 *
+	 * @return bool
+	 */
+	private function can_emoji() {
+		if ( rocket_bypass() || ! $this->options->get( 'emoji' ) ) {
+			return false;
+		}
+
+		return true;
```
Comment (Tabrisrp) [inline inc/Engine/Media/Emojis/EmojisSubscriber.php:98]:
```suggestion
		return ( ! rocket_bypass() || (bool) $this->options->get( 'emoji', 0 ) );
```

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by Tabrisrp** [comment]
@wp-media/qa This PR is ready for QA, acceptance criteria in the OP

**Review by Davidacu — APPROVED** [review]
86% pass rate. 10 Passed, 1 blocked , 1 failed.
Please find the details of the test execution report attached
[testrail-report-13.pdf](https://github.com/wp-media/wp-rocket/files/5031656/testrail-report-13.pdf)

The failure is due to a known [issue ](https://github.com/wp-media/wp-rocket/issues/2293)with heartbeat, but this is out of the scope of this PR.
The blocked case is because we currently don't have a script or plugin that can trigger the heartbeat calls from the frontend, and the issue mentioned above.
Everything else is looking good 👍

**Pull request merged**